File: 0037-sudoku-solver/0037-sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        
        def isValid(row: int, col: int, val: str) -> bool:
            # check row
            for i in range(9):
                if board[row][i] == val or board[i][col] == val:
                    return False
            
            # check 3 * 3
            start_row = (row // 3) * 3
            start_col = (col // 3) * 3
            for i in range(start_row, start_row + 3):
                for j in range(start_col, start_col + 3):
                    if board[i][j] == val:
                        return False
            
            return True
        
        def backtracking() -> bool:
            for i in range(9):
                for j in range(9):
                    if board[i][j].isdigit():
                        continue
                    
                    for char in "123456789":
                        if isValid(i, j, char):
                            board[i][j] = char
                            if backtracking():
                                return True
                            board[i][j] = "."
                    return False
            
            return True
        
        backtracking()
                        

File: 0037-sudoku-solver/0037-sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {
        backtracking(board);
    }

    private boolean backtracking(char[][] board) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                char c = board[i][j];
                if (c != '.') {
                    continue;
                }
                for (char k = '1'; k <= '9'; k++) {
                    if (isValid(board, i, j, k)) {
                        board[i][j] = k;
                        if (backtracking(board)) {
                            return true;
                        }
                        board[i][j] = '.';
                    }
                }
                return false;
            }
        }
        return true;
    }

    private boolean isValid(char[][] board, int i, int j, char k) {
        for (int col = 0; col < 9; col++) {
            if (board[i][col] == k) {
                return false;
            }
        }

        for (int row = 0; row < 9; row++) {
            if (board[row][j] == k) {
                return false;
            }
        }

        int startRow = i / 3 * 3;
        int startCol = j / 3 * 3;
        for (int row = startRow; row < startRow + 3; row++) {
            for (int col = startCol; col < startCol + 3; col++) {
                if (board[row][col] == k) {
                    return false;
                }
            }
        }

        return true;
    }
}